### PR TITLE
Feature/parse payment info and publish in orchestrator

### PR DESCRIPTION
**Pull request by apereverzin**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/BPS-763

### Change description ###
PaymentsPublisher::publishPayments is called after CCD response

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by doncem** [comment]
Looks a lot. Perhaps once you've finely grained the full implementation you can split into 2 (at least)? One for config changes so that they can be done properly (I see bits are missing) and then implementation which uses the config

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/PaymentsProcessorTest.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        caseDetails = CaseDetails.builder().id(20L).build();
+        ArgumentCaptor<PaymentsData> paymentsDataCaptor = ArgumentCaptor.forClass(PaymentsData.class);
+
+        // when
+        paymentsProcessor.processPayments(envelope, caseDetails.getId(), true);
+
+        // then
+        verify(paymentsPublisher).publishPayments(paymentsDataCaptor.capture());
+        PaymentsData paymentsData = paymentsDataCaptor.getValue();
+        assertThat(paymentsData.ccdReference, is(Long.toString(caseDetails.getId())));
```
[the remark this answers is not in the document]
Comment (rkondratowicz) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/PaymentsProcessorTest.java]: we use assertJ (not hamcrest) on this project.
This change should also remove those codacy comments

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/events/CreateExceptionRecord.java [hunk 2]
```diff
@@ -44,21 +44,21 @@ public CreateExceptionRecord(
      * Creates an exception record from given envelope, unless an exception record
      * already exists for this envelope.
      */
-    public void tryCreateFrom(Envelope envelope) {
+    public Long tryCreateFrom(Envelope envelope) {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/events/CreateExceptionRecord.java]: update javadoc

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/servicebus/IPaymentsPublisher.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package uk.gov.hmcts.reform.bulkscan.orchestrator.services.servicebus;
+
+import uk.gov.hmcts.reform.bulkscan.orchestrator.services.servicebus.model.PaymentsData;
+
+public interface IPaymentsPublisher {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/servicebus/IPaymentsPublisher.java]: do we need this interface?

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/servicebus/PaymentsPublisher.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                messageBody,
+                APPLICATION_JSON.toString()
+            );
+
+            queueClient.scheduleMessage(message, Instant.now().plusSeconds(10));
+
+            log.info("Sent message to payments queue. CCD Reference: {}", paymentsData.ccdReference);
+        } catch (Exception ex) {
+            throw new PaymentsPublishingException(
+                "An error occurred when trying to publish payments",
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/servicebus/PaymentsPublisher.java]: we _could_ add ccd ref here.

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/PaymentsProcessorTest.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    @BeforeEach
+    void setUp() {
+        paymentsProcessor = new PaymentsProcessor(paymentsPublisher);
+    }
+
+    @Test
+    void calls_payments_publisher_if_envelope_contains_payments() {
+        // given
+        envelope = SampleData.envelope(1,
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/PaymentsProcessorTest.java]: new line

Comment (apereverzin) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/PaymentsProcessorTest.java]: Sorry

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/servicebus/IPaymentsPublisher.java (hunk 3 above)
Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/servicebus/IPaymentsPublisher.java]: It is used in functional tests similar to IProcessedEnvelopeNotifier

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/events/CreateExceptionRecord.java (hunk 2 above)
Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/events/CreateExceptionRecord.java]: Will do

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/events/EnvelopeHandler.java [hunk 6]
```diff
@@ -13,32 +14,40 @@
...
 
                 break;
             case EXCEPTION:
-            case NEW_APPLICATION:
                 exceptionRecordCreator.tryCreateFrom(envelope);
+                break;
+            case NEW_APPLICATION:
+                Long ccdId = exceptionRecordCreator.tryCreateFrom(envelope);
+
+                paymentsProcessor.processPayments(envelope, ccdId,true);
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/events/EnvelopeHandler.java]: missing space

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/PaymentsProcessor.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+    private final IPaymentsPublisher paymentsPublisher;
+
+    public PaymentsProcessor(
+        IPaymentsPublisher paymentsPublisher
+    ) {
+        this.paymentsPublisher = paymentsPublisher;
+    }
+
+    public void processPayments(Envelope envelope, Long ccdId, boolean isExceptionRecord) {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/PaymentsProcessor.java]: looks like `isExceptionRecord` is always `true`

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/PaymentsProcessorTest.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+@ExtendWith(MockitoExtension.class)
+class PaymentsProcessorTest {
+    @Mock
+    private PaymentsPublisher paymentsPublisher;
+
+    private Envelope envelope;
+
+    private CaseDetails caseDetails;
+
+    private PaymentsProcessor paymentsProcessor;
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/PaymentsProcessorTest.java]: those 3 can be local variables

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/events/EnvelopeHandler.java (hunk 6 above)
Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/events/EnvelopeHandler.java]: Stupid IDE

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/events/CreateExceptionRecord.java [hunk 9]
```diff
@@ -102,6 +106,8 @@ private void createExceptionRecord(Envelope envelope) {
             caseDetails.getId(),
             caseTypeId
         );
+
+        return caseDetails;
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/events/CreateExceptionRecord.java]: looks like only Id is used

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/servicebus/PaymentsPublisher.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        try {
+            String messageBody = objectMapper.writeValueAsString(paymentsData);
+
+            IMessage message = new Message(
+                paymentsData.ccdReference,
+                messageBody,
+                APPLICATION_JSON.toString()
+            );
+
+            queueClient.scheduleMessage(message, Instant.now().plusSeconds(10));
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/servicebus/PaymentsPublisher.java]: is there a reason for the delay?

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/servicebus/PaymentsPublisherTest.java [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+            )
+            .hasCause(exceptionToThrow);
+    }
+
+    private PaymentsData getPaymentsData(boolean isExceptionRecord) {
+        return new PaymentsData(
+                Long.toString(10L),
+                "jurisdiction",
+                "pobox",
+            isExceptionRecord,
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/servicebus/PaymentsPublisherTest.java]: weird formatting

**Comment by apereverzin** [comment]
Created https://github.com/hmcts/bulk-scan-orchestrator/pull/617 to fix the build pipeline.

**Pull request closed without merging**